### PR TITLE
[LoongArch] Minor refinement to monotonic atomic semantics.

### PR DESCRIPTION
**Pull request by tangaac**
Don't use "_db" version AM instructions for LoongArch atomic memory operations with monotonic semantics.

### CODE REVIEW

**Comment by tangaac** [comment]
Need some suggestions to simplify this td file.

**Comment by heiher** [comment]
> Need some suggestions to simplify this td file.

https://github.com/llvm/llvm-project/blob/2f0b4f43fc5c1e7587c4d00daa9cc230df2f8a2d/llvm/lib/Target/AArch64/AArch64InstrAtomics.td#L528-L536
https://github.com/llvm/llvm-project/blob/2f0b4f43fc5c1e7587c4d00daa9cc230df2f8a2d/llvm/lib/Target/AArch64/AArch64InstrFormats.td#L11919-L11951

File: llvm/lib/Target/LoongArch/LoongArchInstrInfo.td [hunk 1]
```diff
@@ -2075,58 +2075,62 @@ multiclass ternary_atomic_op_failure_ord {
 defm atomic_cmp_swap_i32 : ternary_atomic_op_failure_ord;
 defm atomic_cmp_swap_i64 : ternary_atomic_op_failure_ord;
 
+// Atomic operation with the given atomic ordering for word and double word
+multiclass binaray_atomic_op_ord_wd<string inst, string op,
```
Comment (heiher) [inline llvm/lib/Target/LoongArch/LoongArchInstrInfo.td:2079]: `binaray` -> `binary` ?

File: llvm/lib/Target/LoongArch/LoongArchInstrInfo.td [hunk 2]
```diff
@@ -2075,58 +2075,62 @@ multiclass ternary_atomic_op_failure_ord {
...
+
+defm : binaray_atomic_op_ord_wd<"AMSWAP", "atomic_swap", "monotonic">;
+defm : binaray_atomic_op_ord_wd<"AMADD", "atomic_load_add", "monotonic">;
+defm : binaray_atomic_op_ord_wd<"AMAND", "atomic_load_and", "monotonic">;
+defm : binaray_atomic_op_ord_wd<"AMOR", "atomic_load_or", "monotonic">;
+defm : binaray_atomic_op_ord_wd<"AMXOR", "atomic_load_xor", "monotonic">;
+defm : binaray_atomic_op_ord_wd<"AMMIN", "atomic_load_umin", "monotonic", "U">;
+defm : binaray_atomic_op_ord_wd<"AMMAX", "atomic_load_umax", "monotonic", "U">;
+defm : binaray_atomic_op_ord_wd<"AMMIN", "atomic_load_min", "monotonic">;
+defm : binaray_atomic_op_ord_wd<"AMMAX", "atomic_load_max", "monotonic">;
```
Comment (heiher) [inline llvm/lib/Target/LoongArch/LoongArchInstrInfo.td:2105]:
```diff
+multiclass binary_atomic_op_wd<string inst, string op, string signed = ""> {
+  def : Pat<(!cast<PatFrag>(op#"_i32_monotonic") GPR:$rj, GPR:$rk),
+            (!cast<Instruction>(inst#"_W"#signed) GPR:$rk, GPR:$rj)>;
+  def : Pat<(!cast<PatFrag>(op#"_i64_monotonic") GPR:$rj, GPR:$rk),
+            (!cast<Instruction>(inst#"_D"#signed) GPR:$rk, GPR:$rj)>;
+
+  def : Pat<(!cast<PatFrag>(op#"_i32") GPR:$rj, GPR:$rk),
+            (!cast<Instruction>(inst#"__DB_W"#signed) GPR:$rk, GPR:$rj)>;
+  def : Pat<(!cast<PatFrag>(op#"_i64") GPR:$rj, GPR:$rk),
+            (!cast<Instruction>(inst#"__DB_D"#signed) GPR:$rk, GPR:$rj)>;
+}
+

+defm : binary_atomic_op_wd<"AMSWAP", "atomic_swap">;
+defm : binary_atomic_op_wd<"AMADD", "atomic_load_add">;
+defm : binary_atomic_op_wd<"AMAND", "atomic_load_and">;
+defm : binary_atomic_op_wd<"AMOR", "atomic_load_or">;
+defm : binary_atomic_op_wd<"AMXOR", "atomic_load_xor">;
+defm : binary_atomic_op_wd<"AMMIN", "atomic_load_umin", "U">;
+defm : binary_atomic_op_wd<"AMMAX", "atomic_load_umax", "U">;
+defm : binary_atomic_op_wd<"AMMIN", "atomic_load_min">;
+defm : binary_atomic_op_wd<"AMMAX", "atomic_load_max">;
```

How about this?

File: llvm/lib/Target/LoongArch/LoongArchInstrInfo.td (hunk 1 above)
Comment (tangaac) [reply llvm/lib/Target/LoongArch/LoongArchInstrInfo.td:2079]: My mistake.

File: llvm/lib/Target/LoongArch/LoongArchInstrInfo.td (hunk 2 above)
Comment (tangaac) [reply llvm/lib/Target/LoongArch/LoongArchInstrInfo.td:2105]: Very clear, thanks

**Review by SixWeining — APPROVED** [review]
LGTM

**Review by heiher — APPROVED** [review]
LGTM

**Comment by SixWeining** [comment]
About the title and commit message: I think this is not a "fix" but a fine-grained optimization or tuning.

**Comment by SixWeining** [comment]
About `atomic_load_sub`, seems we have something to optimize:

testcase:
```
#include <stdatomic.h>

int *fp2a(int **p) {
  return __atomic_fetch_sub(p, 1, memory_order_relaxed);
}
```

gcc:
```
        addi.w  $r13,$r0,-1                     # 0xffffffffffffffff
        or      $r12,$r4,$r0
        amadd.d $r4,$r13,$r12,0
```

clang:
```
        ori     $a1, $zero, 1
        sub.d   $a2, $zero, $a1
        amadd_db.d      $a1, $a2, $a0
        move    $a0, $a1
```

But this can be done in a separate PR.

**Pull request merged**